### ISSUE TITLE
Add support to repopulate UI

### ISSUE THREAD

**Issue by joiskash** [google/android-fhir#227]
Use case: A questionnaire rendered by the SDK is filed up by a patient. Another phone using the SDK receives the Questionnaire Response. Now a doctor has to :

1. Review the answered questions i.e. read-only mode
2. Enter some answers i.e. edit mode

Originally raised by @kumaakh

**Comment by kumaakh** [comment]
Another interesting case is to save a questionaire in draft state before submitting and then restarting the app/phone and now loading the questionnaire screen again... ideally it should not be required to re answer the questions which were already answered.
Two feature requests here:

1. The SDK should allow bringing the form GUI prefilled with the responses to quietionnaire items and 
2. Allow the form to be rendered in a "View-only" state like described by @joiskash above

**Comment by jingtang10** [comment]
Hi @joiskash and @kumaakh, thanks for filing this issue.

> 1. The SDK should allow bringing the form GUI prefilled with the responses to quietionnaire items and

#189 addresses this issue.

> 2. Allow the form to be rendered in a "View-only" state like described by @joiskash above

I will rename this issue to handle this case.

**Comment by jingtang10** [comment]
This item is only useful after #189 is addressed (there's no point having a read-only mode for an empty questionnaire).

**Comment by jingtang10** [comment]
Sorry @joiskash I just noticed you self-assigned this... I'm not sure which part in @kumaakh's comment you started working on. I'll move this back to To do and can you please self-assign and move #189 to in progress if that's what you're working on? Thanks.

**Comment by mberg** [comment]
Wanted to flag the draft state raised @kumaakh is a really important point.  It would be key to be for a lot of different workflows to be able to jump in and out of a specific form (and between the same type of form between clients) before saving.

**Comment by jingtang10** [comment]
From a UX perspective, do you want the read-only form to be looking the same but all the widgets disabled, or are we talking about a totally different UX?

I think for the sdc library the latter would be out-of-scope. The former makes sense to be a feature of the sdc lib.

**Comment by joiskash** [comment]
> From a UX perspective, do you want the read-only form to be looking the same but all the widgets disabled,

Yes, this is a high priority item for us @kumaakh @Tarun-Bhardwaj

**Comment by mberg** [comment]
I think key idea is SDC can provide a nice way to define how you want to
view data (without the widgets).

**Comment by joiskash** [comment]
Two scopes :
1. [Read only extension in SDC](http://hl7.org/fhir/R4/questionnaire-definitions.html#Questionnaire.item.readOnly) For item level control
2. Initiate the Fragment in only read only for the entire questionnaire and questionnaire response

**Comment by joiskash** [comment]
https://chat.fhir.org/#narrow/stream/179255-questionnaire/topic/instructions.20for.20rendering.20the.20responses.20by.20linkId

**Comment by joiskash** [comment]
https://chat.fhir.org/#narrow/stream/179255-questionnaire/topic/instructions.20for.20rendering.20the.20responses.20by.20linkId

**Comment by fredhersch** [comment]
<img width="412" alt="Review Answers (form-like)" src="https://user-images.githubusercontent.com/34733/134863761-c7f13339-f872-4a6d-a2e0-2c5ec290974c.png">

cc: @joiskash @jingtang10

**Comment by joiskash** [comment]
@nsable7 and I have add a quick discussion. He will be starting off with the following steps

1. Create a new fragment QuestionnaireResponseDisplayFragment
2. According to the image shared by fred, it looks like one single view holder type can be used to display questions and answers. We will call it SimpleQuestionAnswerDisplay
3. Create a new QuestionnaireResponseItemAdapter 
4. Create a new viewModel , we could maybe keep the common code eg: consuming the questionnaire, questionnaire response etc in a super class and then have two sub class view models for data capture and data display.

cc: @jingtang10 please add any comments / suggestions.

**Comment by shelaghm** [comment]
I've simplified the UX design a little since the previous image was uploaded, so attaching a new screenshot (if relevant/helpful)
<img width="412" alt="Review Answers (one edit button)" src="https://user-images.githubusercontent.com/86378921/148419162-107c18e4-a850-461f-a2e2-c08b6500c2a5.png">

**Comment by kumaakh** [comment]
Wouldn't it be better if "Edit" is achieved by tapping on the field which
the user intends to revise ? The generic "Edit" would make the user walk
through the whole wizard  again

**Comment by shelaghm** [comment]
@kumaakh Do you mean having multiple edit buttons, one for each field? I also have a design for that (see below.) Ideally this would enable the user to be taken to the correct place in the questionnaire. 
<img width="412" alt="Review Answers (multiple)" src="https://user-images.githubusercontent.com/86378921/148429626-a0e2462d-63cc-4c56-a2ad-d63f264d63d6.png">

**Comment by kumaakh** [comment]
Yes, I'd prefer the second design,

Also from a visual POV: may be the large edit button can be changed to a
small pencil icon (No text, that's language dependent) without borders.

-Akhil
https://www.linkedin.com/in/akhilk
Ph: +91 9845495104

**Comment by shelaghm** [comment]
From a usability lens, having only an icon is less clear than having the icon combined with text. Especially for an icon like the pencil where it is not necessarily universally known that pencil = edit. If space is a concern, another option is using just the text "edit" or "change" and not the icon.

**Comment by kumaakh** [comment]
Well space is a concern and also the awkwardness of seeing a repeating button on the screen... very distracting.

May be a two step process works: You click the global "Edit" button to show your intent to edit, which starts showing a grey broder arround each question to show that they accept a tap now... and tapping it takes you to the correct step in the questionnaire. in edit mode.

**Comment by mberg** [comment]
With form editing the edit from my perspective needs to be a single
toggle.  We are editing a form and the form can have skip logic and things
like that so there may not even be items you want to edit that are
visible.  I think the second approach is to busy and doesn’t really follow
the pattern I’ve seen in any app.  So I think a single edit button is the
way to go.

**Comment by mberg** [comment]
Just to follow up when you are in edit mode everything needs to be in edit
mode from my perspective.  This is due to skip logic, in form calculations,
etc.  I don’t think you should have to click again to indicate you want to
edit.

**Comment by mberg** [comment]
Sorry the last point is depending on the app there might be quite a bit of
complexity on what we do when we save an edit. If we are editing a health
form it may have a knock on effect on say the patients risk status if we
change a property that relates to that.  We want to do the edit in a single
process.  With partial / incremental edits we wood need to process each
edit as a form edit in case they click out and don’t save the changes

**Comment by kumaakh** [comment]
@mberg this means: In long forms users will have to click "next" multiple times to reach the answer which they are trying to edit, think of the last answer in the form. 

Another option: 
Should we go to full form editing only if the user clicks a calculated field or skipped field (due to skip logic). If they click a simple field it can directly go to that step.

**Comment by mberg** [comment]
Maybe I’m misunderstanding the discussion.  Are we talking about how to
switch to an edit form or something else.  If the form was in edit mode
then we would go back into the data collection mode with data from the
submission prefilling each question.  Apologies if I’m misunderstanding
what this issue is about.

If this is form editing it would be worth to take a look at how odk does
it. That’s the pattern most people are used to

**Comment by kumaakh** [comment]
@mberg yes we are talking about editing the form starting from the review page. I ran through ODK "edit form" functionality and they first land you to the "group view" on edit form action. So the navigation to the correct step becomes much simpler aka hiearchical.  e.g. the "All widgets" sample form in ODK has more than 50 steps grouped into 10 or so groups. So the user first selects one of the 10 groups and then the question which needs to be editted. This way the navigation to the actual item is simpler. 

I realize that this track is not about form editing but about the review feature. So we can defer the "edit" discussion to another track.  I'll review with @joiskash if there is already a track for that or else we can create one.

**Comment by joiskash** [comment]
> Recommendation from discussion 26.08.2021
> 
> Split into two issues (issues to be added), with a priority being on #2
> 
> _1. Support for Questionnaire.item.readOnly:_
> * adds readOnly to the itemControl (e.g. textbox etc)
> * This is in effect the same as applying the readonly attribute as supported by HTML5 (not to be confused with disabled)
> * Lower priority
> 
> _2. SDC feature for rendering Questionnaire/QuestionnaireResponse in a readOnly mode NOT using form itemControls (e.g as name/value pairs in a table)_
> * Could be the final step before submitting a Questionnaire => how do we enable this within the current flow?
> * Load a Questionnaire/QuestionnaireResponse and render for confirming previously entered information
> * In future versions add ability to edit a specific questionnaire item 
> 
> cc: @jingtang10 @joiskash - does this capture it? 

@kumaakh  and @mberg  allowing the ability to edit a specific questionnaire item will be taken care of in future versions.

**Comment by nsabale7** [comment]
Here is a screenshot of the prototype that we have created for this feature as per the design shared above.

![Sample-Questionnaire-Review](https://user-images.githubusercontent.com/82660329/151335558-915da61c-d60a-4318-8906-685499838df3.png)

**Comment by f-odhiambo** [comment]
What's the status of this item @nsabale7 @joiskash?

**Comment by joiskash** [comment]
It is almost done, just waiting to resolve one comment . This should be merged by end of week.
